File: receiver/QrBinary/Protocol/Prng.swift
//
//  Prng.swift
//  QrBinary
//
//  跨语言确定性伪随机数发生器。严格对应 protocol/CONTRACT.md 第 2 节。
//  禁止替换为任何语言内置 RNG，两端必须逐位一致。
//

import Foundation

enum Prng {

    /// lowbias32 finalizer。用于把 blockId 混淆成种子。
    @inline(__always)
    static func mix32(_ x0: UInt32) -> UInt32 {
        var x = x0
        x ^= x >> 16
        x = x &* 0x7FEB_352D
        x ^= x >> 15
        x = x &* 0x846C_A68B
        x ^= x >> 16
        return x
    }

    /// xorshift32 状态推进。0 是不动点，调用方需保证种子非零。
    @inline(__always)
    static func xorshift32(_ s0: UInt32) -> UInt32 {
        var s = s0
        s ^= s &<< 13
        s ^= s >> 17
        s ^= s &<< 5
        return s
    }
}

File: sender/java/src/main/java/org/file/qrcode/protocol/Prng.java
package org.file.qrcode.protocol;

/**
 * 跨语言 PRNG（CONTRACT.md 第 2 节）。
 *
 * 用 int 承载 uint32：所有右移必须写 >>>，乘法自然溢出即模 2^32。
 * 禁止使用 java.util.Random 等语言内置 RNG。
 */
public final class Prng {

    private Prng() {
    }

    /** lowbias32 finalizer，非线性混淆。 */
    public static int mix32(int x) {
        x ^= x >>> 16;
        x *= 0x7FEB352D;
        x ^= x >>> 15;
        x *= 0x846CA68B;
        x ^= x >>> 16;
        return x;
    }

    /** xorshift32 状态推进。0 是不动点，调用方须保证初值非零。 */
    public static int xorshift32(int s) {
        s ^= s << 13;
        s ^= s >>> 17;
        s ^= s << 5;
        return s;
    }
}
